### PR TITLE
Updated to Go 1.18

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
This PR updates Go to 1.18 in all places, including setting the minimum version in `go.mod` files and in all the CI's and containers.

A number of dependencies, most notably the newest Azure SDKs, require Go 1.18 (because of generics) and we will need to update the version of Go used by Dapr to be able to rely on them.

### CODE REVIEW

**Review by berndverst — CHANGES_REQUESTED** [review]
What is causing the current linter failures in this PR? Is there a PR that needs to be merged to address this?

**Comment by ItalyPaleAle** [comment]
Yes this one: https://github.com/dapr/components-contrib/pull/1685

Note that you should not merge this PR (in components-contrib) until the matching PR in dapr/dapr is merged (which is blocked by changing the linter there too). Otherwise dapr/dapr will fail to build if this package requires a newer version of Go.

**Comment by ItalyPaleAle** [comment]
I've also added `go.work` and the respective `go.work.sum` files. New with go 1.18 is that gopls can use the `go.work` file to provide language server capabilities when there are more than one `go.mod` files in a workspace. This fixes a longstanding issue in VS Code where (and other editors that use gopls) where IntelliSense is not working in test apps and other folders that have their own `go.mod`.

File: tests/e2e/pubsub/jetstream/docker-compose.yaml
Comment (daixiang0) [inline tests/e2e/pubsub/jetstream/docker-compose.yaml:12]: Should pin a specific version?

Comment (ItalyPaleAle) [reply tests/e2e/pubsub/jetstream/docker-compose.yaml:12]: This pins to 1.18 but allows patch updates.

Comment (berndverst) [reply tests/e2e/pubsub/jetstream/docker-compose.yaml:12]: @daixiang0 this should be ok -- patch version changes are fine

**Review by berndverst — APPROVED** [review]

**Pull request merged**
